### ISSUE TITLE
sorbet autocorrect not suggested for T.nilable(Type) return values when returns(Type).

### ISSUE THREAD

**Issue by lgo** [sorbet/sorbet#1420]
#### Input

sorbet does not seem to recommend an autocorrection for calling `T.must` on a `T.nilable(Type)` when the method signature has `returns(Type)`.

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aextend%20T%3A%3ASig%0A%0A%23%20Does%20not%20suggest%20the%20error%20can%20be%20auto-corrected.%0A%0Asig%20%7Breturns(String)%7D%0Adef%20foo%0A%20%20T.let(%22bar%22%2C%20T.nilable(String))%0Aend%0A%0A%23%20Unlike%20the%20below%2C%20where%20T.must%20is%20suggested.%0Aa%20%3D%20T.let(%22bar%22%2C%20T.nilable(String))%0Aa%20%2B%20%22bar%22)

```ruby
# typed: true

extend T::Sig

# Does not suggest the error can be auto-corrected.

sig {returns(String)}
def foo
  T.let("bar", T.nilable(String))
end

# Unlike the below, where T.must is suggested.
a = T.let("bar", T.nilable(String))
a + "bar"
```

I might expect that sorbet recommends an auto-correction in the first case to add a `T.must`.

**Comment by jez** [comment]
We reverted because we overlooked some edge cases. Marking this as good first issue for anyone who wants to resurrect it.

**Comment by jez** [comment]
I think this is now mostly fixed by way of #4144
